api/src/main/java/io/opentelemetry/common/AttributeKeyImpl.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.common;
+
+import java.util.List;
+
+@SuppressWarnings("rawtypes")
+public abstract class AttributeKeyImpl<T> implements AttributeKey<T> {
+  private final String key;
+
+  AttributeKeyImpl(String key) {
+    this.key = key;
+  }
+
+  @Override
+  public String get() {
+    return key;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof AttributeKeyImpl)) {
+      return false;
+    }
+
+    AttributeKeyImpl<?> that = (AttributeKeyImpl<?>) o;
+
+    return key != null ? key.equals(that.key) : that.key == null;
+  }
+
+  @Override
+  public int hashCode() {
+    return key != null ? key.hashCode() : 0;
+  }
+
+  @Override
+  public String toString() {
+    return "AttributeKeyImpl{" + "key='" + key + '\'' + '}';
```

Comment:
I'm on the fence. It will certainly read better, but I wonder if it could be misleading in tests if the failure happens when you're trying to compare to a raw String, and the toStrings would look identical.

Reply:
Admittedly I'm biased because instrumentation tests are groovy and would be able to use the key itself in assertions instead of calling a method. This is not a strong enough reason to pick one or another though so ok with either 😃

Reply:
wait...spock uses the toString to do assertions??? That sounds like a terrible idea.